google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQuery.java

```diff
@@ -440,6 +443,72 @@ public static QueryResultsOption maxWaitTime(long maxWaitTime) {
     }
   }
 
+  class QueryOption implements Serializable {
+    private static final long serialVersionUID = 6206193419355824689L;
+
+    private Object option;
+
+    private QueryOption(Object option) {
+      this.option = option;
+    }
+
+    public QueryResultsOption getQueryResultsOption() {
+      return option instanceof QueryResultsOption ? (QueryResultsOption) option : null;
+    }
+
+    public RetryOption getRetryOption() {
+      return option instanceof RetryOption ? (RetryOption) option : null;
+    }
+
+    public static QueryResultsOption[] filterQueryResutlsOptions(QueryOption... options) {
```

Comment:
"Results"?

Why arrays? I'd think lists are more idiomatic?

Reply:
Yes, we should use collections as return types instead of arrays if possible. In this particular case arrays are used because our manual clients use Options in a form of vararg arguments (someMethod(QueryResultsOption... options), so the output of this method is most likely to be passed to a method, which accepts a vararg argument of QueryResultsOption, and we can pass an array as a vararg argument (and it would work as expected). If it was a list we would have to convert it to an array first. So I decided to return array directly to avoid unnecessary conversions.

Reply:
ah, sgtm